File: fe/poisson.f90
!--------------------------------------------------------------------------!
program poisson                                                            !
!--------------------------------------------------------------------------!

use sigma
use fe

implicit none

    type(triangle_mesh) :: mesh
    type(cs_graph), pointer :: g
    type(csr_matrix) :: A, B
    type(bicgstab_solver) :: bcg
    type(sparse_ldu_solver) :: ildu
    real(dp), allocatable :: u(:), f(:), z(:)
    integer :: i, j, k, d, nn
    integer, allocatable :: nodes(:)
    logical :: success

    call init_seed()

    success = mesh%read_mesh("meshes/example.2")

    if (.not. success) then
        print *, "Oh no! Failed to read mesh!"
        call exit(1)
    endif

    nn = mesh%num_nodes

    print *, "    + Successfully read triangular mesh."
    print *, "        - Number of nodes:    ", mesh%num_nodes
    print *, "        - Number of edges:    ", mesh%num_edges
    print *, "        - Number of triangles:", mesh%num_triangles

    allocate(g)
    call build_connectivity_graph(g, mesh)
    call A%set_dimensions(nn, nn)
    call B%set_dimensions(nn, nn)
    call A%set_graph(g)
    call B%set_graph(g)
    call A%zero()
    call B%zero()

    call fill_poisson_stiffness_matrix(A, mesh)
    call fill_p1_mass_matrix(B, mesh)

    print *, "    + Built finite element stiffness and mass matrices."
    print *, "    + Generating random right-hand side for Poisson problem."

    allocate(u(nn), f(nn), z(nn))
    u = 0.0_dp
    f = 0.0_dp

    call random_number(z)
    z = z - 0.5_dp

    call B%matvec(z, f)

    d = g%get_max_degree()
    allocate(nodes(d))

    do i = 1, nn
        if (mesh%node_boundary(i) /= 0) then
            call g%get_neighbors(nodes, i)
            d = g%get_degree(i)

            do k = 1, d
                j = nodes(k)
                call A%set(i, j, 0.0_dp)
            enddo

            call A%set(i, i, 1.0_dp)

            f(i) = 0.0_dp
        endif
    enddo


    call bcg%setup(A)

    call bcg%solve(A, u, f)

    print *, "    + Done solving Poisson problem."
    print *, "      BiCG-Stab solver iterations:", bcg%iterations
    print *, "    + Writing output to file u.txt."


    open(file = "u.txt", unit = 10)
    do i = 1, nn
        write(10, *) u(i)
    enddo
    close(10)


    call bcg%destroy()
    call B%destroy()
    call A%destroy()
    call g%destroy()
    deallocate(g)
    deallocate(u, f, z)

end program poisson

File: python-interface/poisson.c

#include "triangle_meshes.h"
#include "glue.h"
#include <stdio.h>
#include <stdlib.h>

#define max(a, b) (a < b) ? b : a;
#define min(a, b) (a < b) ? a : b;

int main(int argc, char **argv) {
    triangle_mesh mesh;
    bool success = read_mesh(&mesh, "meshes/example.2");

    if (success) {
        printf("Successfully read mesh! Hoorah!\n");
    } else {
        printf("Womp womp, failed to read mesh.\n");
        return 1;
    }

    double *u = calloc(mesh.num_nodes, sizeof(double));

    for (int i = 0; i < mesh.num_nodes; i++) u[i] = 0.0;

    driver(mesh.num_nodes, mesh.num_edges, mesh.num_triangles,
            mesh.x, mesh.y, mesh.edges, mesh.triangles, mesh.node_boundary, u);

    double mi = 0.0, ma = 0.0;

    for (int i = 0; i < mesh.num_nodes; i++) {
        mi = min(mi, u[i]);
        ma = max(ma, u[i]);
    }

    printf("%g %g\n", mi, ma);

    free(u);
    destroy_mesh(&mesh);

    return 0;
}
